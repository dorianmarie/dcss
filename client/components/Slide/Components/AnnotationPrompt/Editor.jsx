import React from 'react';
import PropTypes from 'prop-types';
import DataHeader from '@components/Slide/Components/DataHeader';
import { Button, Dropdown, Form, Table } from '@components/UI';
import Identity from '@utils/Identity';
import { type } from './meta';

const AnnotationPromptFormatted = ({
  title,
  prompt,
  header,
  isConditional
}) => {
  return (
    <Table celled striped>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell colSpan="2">{title}</Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      <Table.Body>
        <Table.Row>
          <Table.Cell collapsing>Prompt:</Table.Cell>
          <Table.Cell>{prompt}</Table.Cell>
        </Table.Row>
        {header ? (
          <Table.Row>
            <Table.Cell collapsing>Data header:</Table.Cell>
            <Table.Cell>{header}</Table.Cell>
          </Table.Row>
        ) : null}
        {isConditional ? (
          <Table.Row>
            <Table.Cell colSpan={2}>
              This prompt appears conditionally.
            </Table.Cell>
          </Table.Row>
        ) : null}
      </Table.Body>
    </Table>
  );
};

AnnotationPromptFormatted.propTypes = {
  title: PropTypes.string,
  prompt: PropTypes.string,
  header: PropTypes.string,
  isConditional: PropTypes.bool
};

class AnnotationPromptEditor extends React.Component {
  constructor(props) {
    super(props);

    const { header, question, prompts, responseId } = props.value;

    this.state = {
      components: null,
      header,
      question,
      prompts,
      responseId
    };
    this.timeout = null;
    this.delayedUpdateState = this.delayedUpdateState.bind(this);
    this.updateState = this.updateState.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  async componentDidMount() {
    const { scenario } = this.props;
    // TODO: move to async action
    const { components } = await (await fetch(
      `/api/scenarios/${scenario.id}/slides/prompt-components`
    )).json();

    if (!this.hasUnmounted) {
      this.setState({ components });
    }
  }

  componentWillUnmount() {
    this.hasUnmounted = true;

    clearTimeout(this.timeout);

    const { header, prompts, question, responseId } = this.props.value;
    const lastProps = {
      header,
      prompts,
      question,
      responseId
    };

    if (Identity.key(this.state) !== Identity.key(lastProps)) {
      this.updateState();
    }
  }

  delayedUpdateState() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
    this.timeout = setTimeout(this.updateState, 500);
  }

  updateState() {
    const { header, question, responseId, prompts } = this.state;
    this.props.onChange({
      ...this.props.value,
      header,
      question,
      responseId,
      prompts,
      type
    });
  }

  onChange(event, { name, value }) {
    this.setState({ [name]: value }, this.delayedUpdateState);
  }

  render() {
    const {
      slideId,
      value: { header = '', prompts = [], question = '' }
    } = this.props;
    const { components } = this.state;
    const { onChange, updateState } = this;

    if (!components) {
      return null;
    }

    const reviewables = components.reduce((accum, component, key) => {
      const { header, isConditional, prompt, responseId, slide } = component;

      // Don't include empty/incomplete prompts
      // Don't include prompts from THIS slide
      // Don't include components with `disableEmbed: true`
      if (
        !responseId ||
        slideId === component.slide.id ||
        component.disableEmbed
      ) {
        return accum;
      }

      const quotedSlideTitle = slide.title ? ` "${slide.title}"` : ``;
      const title = `Slide #${slide.slide_number} ${quotedSlideTitle}`;
      const content = (
        <AnnotationPromptFormatted
          title={title}
          prompt={prompt}
          header={header}
          isConditional={isConditional}
        />
      );

      const text = `Slide #${slide.slide_number}: "${prompt}"`;
      accum.push({
        key: `recall-response-${key}`,
        text,
        content,
        value: responseId
      });

      return accum;
    }, []);

    if (!reviewables.length) {
      reviewables.push({
        key: 'recall-response-missing',
        text: 'No participant responses available',
        value: null
      });
    }

    const defaultValue = prompts || [];
    const selectAnnotationDropdown = (
      <Dropdown
        fluid
        multiple
        selection
        name="prompts"
        value={defaultValue}
        onChange={(event, { name, value }) => {
          onChange(event, {
            name,
            value: value.filter(Boolean)
          });
        }}
        options={reviewables}
      />
    );

    const displayParticipantAnnotationLabel = `Select the prompts to be reviewed by the participant:`;

    const selectAnnotationPromptsFormFieldLabelled = (
      <Form.Field.Labelled
        label={displayParticipantAnnotationLabel}
        content={selectAnnotationDropdown}
      />
    );

    // const selectedPrompts = prompts.length
    //   ? reviewables.filter(reviewable => prompts.includes(reviewable.responseId))
    //   : [];

    // const selectedPromptContent = selectedPrompts.length ? (
    //   <Fragment>
    //     {selectedPrompts.map(selected => selected.content)}
    //   </Fragment>
    // ) : null

    const selectAllPromptsLabel = 'Select all prompts';

    const selectAllPrompts = (
      <Button
        fluid
        icon="arrow left"
        labelPosition="left"
        className="icon-primary"
        content={selectAllPromptsLabel}
        onClick={() => {
          const name = 'prompts';
          const value = reviewables.map(prompt => prompt.value).filter(Boolean);
          onChange(
            {},
            {
              name,
              value
            }
          );
        }}
      />
    );

    const selectAllPromptsFormFieldLabelled = (
      <Form.Field.Labelled label="&nbsp;" content={selectAllPrompts} />
    );

    const questionAriaLabel = 'Question to ask about these prompts:';

    return (
      <Form>
        <Form.TextArea
          required
          name="question"
          label={questionAriaLabel}
          aria-label={questionAriaLabel}
          rows={1}
          defaultValue={question}
          onChange={onChange}
          onBlur={updateState}
        />

        <Form.Group widths="equal">
          {selectAnnotationPromptsFormFieldLabelled}
          {selectAllPromptsFormFieldLabelled}
        </Form.Group>

        <DataHeader content={header} onChange={onChange} onBlur={updateState} />
      </Form>
    );
  }
}

AnnotationPromptEditor.propTypes = {
  onChange: PropTypes.func.isRequired,
  scenario: PropTypes.object,
  slideId: PropTypes.any,
  slideIndex: PropTypes.any,
  value: PropTypes.shape({
    id: PropTypes.string,
    header: PropTypes.string,
    prompts: PropTypes.array,
    question: PropTypes.string,
    responseId: PropTypes.string,
    type: PropTypes.oneOf([type])
  })
};

export default AnnotationPromptEditor;
