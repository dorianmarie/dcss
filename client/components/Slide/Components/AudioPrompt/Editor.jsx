import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Identity from '@utils/Identity';
import { Container, Form, Message } from '@components/UI';
import { type } from './meta';
// import AgentSelector from '@components/Slide/Components/AgentSelector';
import DataHeader from '@components/Slide/Components/DataHeader';
import ResponseRecall from '@components/Slide/Components/ResponseRecall/Editor';
import './AudioPrompt.css';
import '@components/Slide/SlideEditor/SlideEditor.css';

class AudioPromptEditor extends Component {
  constructor(props) {
    super(props);
    const {
      // agent = null,
      header = '',
      prompt = '',
      recallId = '',
      responseId
    } = props.value;

    this.state = {
      // agent,
      header,
      prompt,
      recallId,
      responseId
    };
    this.onChange = this.onChange.bind(this);
    this.onRecallChange = this.onRecallChange.bind(this);
    this.updateState = this.updateState.bind(this);
    this.delayedUpdateState = this.delayedUpdateState.bind(this);
    this.timeout = null;
  }

  componentWillUnmount() {
    clearTimeout(this.timeout);

    const {
      // agent,
      header, prompt, recallId, responseId } = this.props.value;

    const lastProps = {
      // agent,
      header,
      prompt,
      recallId,
      responseId
    };

    if (Identity.key(this.state) !== Identity.key(lastProps)) {
      this.updateState();
    }
  }

  delayedUpdateState() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }

    this.timeout = setTimeout(this.updateState, 500);
  }

  updateState() {
    const {
      // agent,
      header, prompt, recallId, responseId } = this.state;

    this.props.onChange({
      // agent,
      header,
      prompt,
      recallId,
      responseId,
      type
    });
  }

  onChange(event, { name, value }) {
    this.setState({ [name]: value }, this.delayedUpdateState);
  }

  onRecallChange({ recallId }) {
    this.setState({ recallId }, this.updateState);
  }

  render() {
    const {
      // agent,
      header, prompt, recallId } = this.state;
    const { scenario, slideIndex } = this.props;
    const { onChange, onRecallChange, updateState } = this;

          // <AgentSelector
          //   label="Optional AI agent:"
          //   agent={agent}
          //   type={type}
          //   onChange={onChange}
          // />

    return (
      <Form>
        <Container fluid>
          <ResponseRecall
            isEmbedded={true}
            value={{ recallId }}
            scenario={scenario}
            slideIndex={slideIndex}
            onChange={onRecallChange}
          />

          <Form.TextArea
            label="Optional prompt to display before the audio recording controls:"
            name="prompt"
            rows={1}
            value={prompt}
            onChange={onChange}
            onBlur={updateState}
          />

          <DataHeader
            content={header}
            onChange={onChange}
            onBlur={updateState}
          />

          <Message
            icon="warning sign"
            color="orange"
            content="This component will fallback to a text input prompt when the participant's browser or device does not support audio recording."
          />
        </Container>
      </Form>
    );
  }
}

AudioPromptEditor.propTypes = {
  onChange: PropTypes.func.isRequired,
  scenario: PropTypes.object,
  slideIndex: PropTypes.any,
  value: PropTypes.shape({
    id: PropTypes.string,
    header: PropTypes.string,
    prompt: PropTypes.string,
    recallId: PropTypes.string,
    required: PropTypes.bool,
    responseId: PropTypes.string,
    type: PropTypes.oneOf([type, 'AudioResponse'])
  })
};

export default AudioPromptEditor;
