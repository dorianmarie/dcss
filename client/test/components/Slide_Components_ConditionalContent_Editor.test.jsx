/** @TEMPLATE: BEGIN **/
import React from 'react';
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useLayoutEffect: jest.requireActual('react').useEffect,
}));

import {
  fetchImplementation,
  mounter,
  reduxer,
  serialize,
  snapshotter,
  state,
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';

import {
  act,
  fireEvent,
  prettyDOM,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

async function waitForPopper() {
  // Popper update() - https://github.com/popperjs/react-popper/issues/350
  await act(async () => await null);
}

/** @TEMPLATE: END **/

/** @GENERATED: BEGIN **/

import { GET_SCENARIO_PROMPT_COMPONENTS_SUCCESS } from '../../actions/types';
import * as scenarioActions from '../../actions/scenario';
jest.mock('../../actions/scenario');

let prompt;
let prompts;
let scenario;
let slideIndex;
let value;

import Editor from '../../components/Slide/Components/ConditionalContent/Editor.jsx';
/** @GENERATED: END **/

/** @TEMPLATE: BEGIN **/
const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;
/** @TEMPLATE: END **/

beforeAll(() => {
  /** @TEMPLATE: BEGIN **/
  (window || global).fetch = jest.fn();
  /** @TEMPLATE: END **/
});

afterAll(() => {
  /** @TEMPLATE: BEGIN **/
  jest.restoreAllMocks();
  /** @TEMPLATE: END **/
});

beforeEach(() => {
  /** @TEMPLATE: BEGIN **/
  jest.useFakeTimers();
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
  /** @TEMPLATE: END **/

  /** @GENERATED: BEGIN **/

  scenario = {
    author: {
      id: 999,
      username: 'super',
      personalname: 'Super User',
      email: 'super@email.com',
      is_anonymous: false,
      roles: ['participant', 'super_admin', 'facilitator', 'researcher'],
      is_super: true,
    },
    categories: [],
    consent: { id: 57, prose: '' },
    description: "This is the description of 'A Multiplayer Scenario'",
    finish: {
      id: 1,
      title: '',
      components: [
        { html: '<h2>Thanks for participating!</h2>', type: 'Text' },
      ],
      is_finish: true,
    },
    lock: {
      scenario_id: 42,
      user_id: 999,
      created_at: '2020-02-31T23:54:19.934Z',
      ended_at: null,
    },
    slides: [
      {
        id: 1,
        title: '',
        components: [
          { html: '<h2>Thanks for participating!</h2>', type: 'Text' },
        ],
        is_finish: true,
      },
      {
        id: 2,
        title: '',
        components: [
          {
            id: 'b7e7a3f1-eb4e-4afa-8569-eb6677358c9e',
            html: '<p>paragraph</p>',
            type: 'Text',
          },
          {
            agent: null,
            id: 'aede9380-c7a3-4ef7-add7-838fd5ec854f',
            type: 'TextResponse',
            header: 'TextResponse-1',
            prompt: '',
            timeout: 0,
            recallId: '',
            required: true,
            responseId: 'be99fe9b-fa0d-4ab7-8541-1bfd1ef0bf11',
            placeholder: '',
          },
          {
            id: 'f96ac6de-ac6b-4e06-bd97-d97e12fe72c1',
            html: '<p>?</p>',
            type: 'Text',
          },
        ],
        is_finish: false,
      },
    ],
    status: 1,
    title: 'Multiplayer Scenario 2',
    users: [
      {
        id: 999,
        email: 'super@email.com',
        username: 'super',
        personalname: 'Super User',
        roles: ['super'],
        is_super: true,
        is_author: true,
        is_reviewer: false,
      },
    ],
    id: 42,
    created_at: '2020-08-31T17:50:28.089Z',
    updated_at: null,
    deleted_at: null,
    labels: ['a', 'b'],
    personas: [
      {
        id: 1,
        name: 'Participant',
        description:
          'The default user participating in a single person scenario.',
        color: '#FFFFFF',
        created_at: '2020-12-01T15:49:04.962Z',
        updated_at: null,
        deleted_at: null,
        author_id: 3,
        is_read_only: true,
        is_shared: true,
      },
    ],
  };
  slideIndex = 0;
  prompt = {
    agent: { id: 1, rules: Array(0) },
    header: 'TextResponse-0',
    id: '80b92e32-1694-4a38-b3d1-a090f68ff7de',
    index: 0,
    persona: null,
    placeholder: '',
    prompt: 'Hello this is my optional prompt. ',
    recallId: '',
    required: true,
    responseId: '16b011f7-62cc-47bb-a33f-4cb292d285bd',
    slide: {},
    timeout: 0,
    type: 'TextResponse',
  };

  prompts = [prompt];

  value = {
    agent: null,
    disableRequireCheckbox: true,
    header: '',
    id: 'XYZ',
    rules: [
      {
        operator: null,
        value: null,
      },
    ],
    persona: null,
    recallId: 'xyz-recallId',
    required: false,
    responseId: 'xyz-responseId',
    html: '<p></p>',
    type: 'ConditionalContent',
  };

  scenarioActions.getScenarioPromptComponents.mockImplementation(
    () => async (dispatch) => {
      dispatch({ type: GET_SCENARIO_PROMPT_COMPONENTS_SUCCESS, prompts });
      return prompts;
    }
  );

  /** @GENERATED: END **/

  /** @TEMPLATE: BEGIN **/
  commonProps = {};
  commonState = JSON.parse(JSON.stringify(original));
  /** @TEMPLATE: END **/
});

afterEach(() => {
  /** @TEMPLATE: BEGIN **/
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
  jest.resetAllMocks();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
  /** @TEMPLATE: END **/
});

test('Editor', () => {
  expect(Editor).toBeDefined();
});

/** @GENERATED: BEGIN **/
test('Render 1 1', async (done) => {
  const Component = Editor;
  const props = {
    ...commonProps,
    type: 'ConditionalContent',
    value,
    scenario,
    onChange: jest.fn(),
  };

  const state = {
    ...commonState,
  };

  const ConnectedRoutedComponent = reduxer(Component, props, state);

  await render(<ConnectedRoutedComponent {...props} />);
  expect(serialize()).toMatchSnapshot();

  done();
});
/** @GENERATED: END **/

