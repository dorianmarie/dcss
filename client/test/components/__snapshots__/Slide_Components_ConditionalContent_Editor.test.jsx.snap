// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `
"<body>
  <div
    id=\\"root\\"
  />
  <div>
    <form
      class=\\"ui form\\"
    >
      <div
        class=\\"ui fluid container\\"
      >
        <div
          class=\\"ui grid\\"
        >
          <div
            class=\\"row\\"
          >
            <div
              class=\\"column\\"
            >
              <div
                class=\\"field\\"
              >
                <label>
                  Select an AI agent:
                </label>
                <div
                  aria-expanded=\\"false\\"
                  aria-label=\\"Select an agent\\"
                  class=\\"ui fluid search selection scrolling dropdown\\"
                  role=\\"combobox\\"
                >
                  <input
                    aria-autocomplete=\\"list\\"
                    autocomplete=\\"off\\"
                    class=\\"search\\"
                    tabindex=\\"0\\"
                    type=\\"text\\"
                    value=\\"\\"
                  />
                  <div
                    aria-atomic=\\"true\\"
                    aria-live=\\"polite\\"
                    class=\\"divider default text\\"
                    role=\\"alert\\"
                  >
                    Select an agent
                  </div>
                  <i
                    aria-hidden=\\"true\\"
                    class=\\"dropdown icon\\"
                  />
                  <div
                    class=\\"menu transition\\"
                    role=\\"listbox\\"
                  >
                    <div
                      aria-checked=\\"false\\"
                      aria-selected=\\"true\\"
                      class=\\"selected item\\"
                      role=\\"option\\"
                      style=\\"pointer-events: all;\\"
                    >
                      <span
                        class=\\"text\\"
                      >
                        No selection
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class=\\"ui segment\\"
              >
                <p
                  class=\\"cce__paragraph\\"
                  tabindex=\\"0\\"
                >
                  Create rules that are used to determine if the content below will be displayed to the participant.
                </p>
                <p
                  class=\\"cce__paragraph\\"
                  tabindex=\\"0\\"
                >
                  For every rule, 
                  <code>
                    X
                  </code>
                   is the sum of affirmative responses from the agent, and 
                  <code>
                    Y
                  </code>
                   is your comparison value.
                </p>
              </div>
              <table
                class=\\"ui definition striped unstackable table\\"
              >
                <thead
                  class=\\"\\"
                >
                  <tr
                    class=\\"\\"
                  >
                    <th
                      class=\\"mpr__thead-background\\"
                    />
                    <th
                      class=\\"\\"
                    >
                      Operation
                    </th>
                    <th
                      class=\\"cce__cell-constraint\\"
                    >
                      Y value
                    </th>
                    <th
                      class=\\"cce__cell-constraint\\"
                    >
                      Evaluates as...
                    </th>
                  </tr>
                </thead>
                <tbody
                  class=\\"\\"
                >
                  <tr
                    class=\\"mpr__cursor-grab\\"
                  >
                    <td
                      class=\\"collapsing\\"
                    >
                      <div
                        class=\\"ui borderless icon em__height menu\\"
                      >
                        <a
                          aria-label=\\"Save rule\\"
                          class=\\"item\\"
                          tabindex=\\"0\\"
                        >
                          <i
                            aria-hidden=\\"true\\"
                            class=\\"save outline icon\\"
                          />
                        </a>
                        <a
                          aria-label=\\"Delete this rule\\"
                          class=\\"item\\"
                          tabindex=\\"0\\"
                        >
                          <i
                            aria-hidden=\\"true\\"
                            class=\\"trash alternate outline icon\\"
                          />
                        </a>
                      </div>
                    </td>
                    <td
                      class=\\"\\"
                    >
                      <div
                        aria-expanded=\\"false\\"
                        aria-label=\\"Select the operator for expression 1\\"
                        class=\\"ui fluid search selection dropdown\\"
                        index=\\"0\\"
                        name=\\"operator\\"
                        role=\\"combobox\\"
                      >
                        <input
                          aria-autocomplete=\\"list\\"
                          autocomplete=\\"off\\"
                          class=\\"search\\"
                          tabindex=\\"0\\"
                          type=\\"text\\"
                          value=\\"\\"
                        />
                        <i
                          aria-hidden=\\"true\\"
                          class=\\"dropdown icon\\"
                        />
                        <div
                          class=\\"menu transition\\"
                          role=\\"listbox\\"
                        >
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"true\\"
                            class=\\"selected item\\"
                            id=\\"\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          />
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"false\\"
                            class=\\"item\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          >
                            <table
                              class=\\"ui table text\\"
                            >
                              <thead
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <th
                                    class=\\"\\"
                                  >
                                    <code>
                                      X == Y
                                    </code>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <td
                                    class=\\"\\"
                                  >
                                    X equals Y
                                    :
                                    <br />
                                    Use this operation to determine if \\"X\\", which is the sum of affirmative responses from this agent, is equal to \\"Y\\", which is your comparison value.
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"false\\"
                            class=\\"item\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          >
                            <table
                              class=\\"ui table text\\"
                            >
                              <thead
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <th
                                    class=\\"\\"
                                  >
                                    <code>
                                      X != Y
                                    </code>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <td
                                    class=\\"\\"
                                  >
                                    X does not equal Y
                                    :
                                    <br />
                                    Use this operation to determine if \\"X\\", which is the sum of affirmative responses from this agent, is not equal to \\"Y\\", which is your comparison value.
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"false\\"
                            class=\\"item\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          >
                            <table
                              class=\\"ui table text\\"
                            >
                              <thead
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <th
                                    class=\\"\\"
                                  >
                                    <code>
                                      X &gt; Y
                                    </code>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <td
                                    class=\\"\\"
                                  >
                                    X is greater than Y
                                    :
                                    <br />
                                    Use this operation to determine if \\"X\\", which is the sum of affirmative responses from this agent, is greater than \\"Y\\", which is your comparison value.
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"false\\"
                            class=\\"item\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          >
                            <table
                              class=\\"ui table text\\"
                            >
                              <thead
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <th
                                    class=\\"\\"
                                  >
                                    <code>
                                      X &gt;= Y
                                    </code>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <td
                                    class=\\"\\"
                                  >
                                    X is greater than or equal to Y
                                    :
                                    <br />
                                    Use this operation to determine if \\"X\\", which is the sum of affirmative responses from this agent, is greater than or equal to \\"Y\\", which is your comparison value.
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"false\\"
                            class=\\"item\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          >
                            <table
                              class=\\"ui table text\\"
                            >
                              <thead
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <th
                                    class=\\"\\"
                                  >
                                    <code>
                                      X &lt; Y
                                    </code>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <td
                                    class=\\"\\"
                                  >
                                    X is less than Y
                                    :
                                    <br />
                                    Use this operation to determine if \\"X\\", which is the sum of affirmative responses from this agent, is less than \\"Y\\", which is your comparison value.
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                          <div
                            aria-checked=\\"false\\"
                            aria-selected=\\"false\\"
                            class=\\"item\\"
                            role=\\"option\\"
                            style=\\"pointer-events: all;\\"
                          >
                            <table
                              class=\\"ui table text\\"
                            >
                              <thead
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <th
                                    class=\\"\\"
                                  >
                                    <code>
                                      X &lt;= Y
                                    </code>
                                  </th>
                                </tr>
                              </thead>
                              <tbody
                                class=\\"\\"
                              >
                                <tr
                                  class=\\"\\"
                                >
                                  <td
                                    class=\\"\\"
                                  >
                                    X is less than or equal to Y
                                    :
                                    <br />
                                    Use this operation to determine if \\"X\\", which is the sum of affirmative responses from this agent, is less than or equal to \\"Y\\", which is your comparison value.
                                  </td>
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                      </div>
                    </td>
                    <td
                      class=\\"\\"
                    >
                      <div
                        class=\\"ui fluid input\\"
                        index=\\"0\\"
                        options=\\"[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]\\"
                      >
                        <input
                          aria-label=\\"Enter the value for expression 1\\"
                          autocomplete=\\"off\\"
                          name=\\"value\\"
                          type=\\"text\\"
                          value=\\"\\"
                        />
                      </div>
                    </td>
                    <td
                      class=\\"\\"
                    >
                      Select an operator to create a rule
                    </td>
                  </tr>
                </tbody>
                <tfoot
                  class=\\"full-width\\"
                >
                  <tr
                    class=\\"\\"
                  >
                    <th
                      class=\\"\\"
                    />
                    <th
                      class=\\"\\"
                      colspan=\\"3\\"
                    >
                      <div
                        class=\\"ui borderless right floated menu\\"
                      >
                        <a
                          aria-label=\\"Add another slide choice\\"
                          class=\\"icon item\\"
                          tabindex=\\"0\\"
                        >
                          <i
                            class=\\"large icons em__icon-group-margin\\"
                          >
                            <i
                              aria-hidden=\\"true\\"
                              class=\\"hand pointer outline icon\\"
                            />
                            <i
                              aria-hidden=\\"true\\"
                              class=\\"green add top right corner icon\\"
                            />
                          </i>
                          Add a rule
                        </a>
                      </div>
                    </th>
                  </tr>
                </tfoot>
              </table>
              <textarea />
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</body>"
`;
